File: python/98-Validate-Binary-Search-Tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


import math
from typing import Optional

# region Recursive Traversal w/ Valid Range
# Time O(n)
# Space O(n)


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.validate(root, -math.inf, math.inf)

    def validate(self, root, curMin, curMax) -> bool:
        if not root:
            return True

        if root.val >= curMax or root.val <= curMin:
            return False

        return self.validate(root.left, curMin, root.val) and self.validate(
            root.right, root.val, curMax
        )


# endregion


# region Iterative Traversal w/ Valid Range
# Space O(n)
# Time O(n)
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        stack = []
        stack.append([root, -math.inf, math.inf])

        while stack:
            curNodeAndBounds = stack.pop()
            curNode = curNodeAndBounds[0]
            if curNode.val <= curNodeAndBounds[1] or curNode.val >= curNodeAndBounds[2]:
                return False
            if curNode.left:
                stack.append([curNode.left, curNodeAndBounds[1], curNode.val])
            if curNode.right:
                stack.append([curNode.right, curNode.val, curNodeAndBounds[2]])
        return True


# endregion


# region Recursive In Order Traversal
# Space O(n) -- Worst case when there's a "bad" elem in the rightmost leaf
# Time O(n)
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        self.prev = -math.inf

        def inOrder(root) -> bool:
            if not root:
                return True

            if not inOrder(root.left):
                return False

            if root.val <= self.prev:
                return False
            self.prev = root.val
            return inOrder(root.right)

        return inOrder(root)


# endregion

File: c#/98-Validate-Binary-Search-Tree.cs
#region DFS
/// <summary>
/// 98. Validate Binary Search Tree
/// https://leetcode.com/problems/validate-binary-search-tree/
///
/// Given a binary tree, determine if it is a valid binary search tree (BST).
///
/// Assume a BST is defined as follows:
/// The left subtree of a node contains only nodes with keys less than the node's key.
/// The right subtree of a node contains only nodes with keys greater than the node's key.
/// Both the left and right subtrees must also be binary search trees.
///
/// Example 1:
/// Input: root = [2,1,3]
/// Output: true
///
/// Example 2:
/// Input: root = [5,1,4,null,null,3,6]
/// Output: false
///
/// Example 3:
/// Input: root = [1,1]
/// Output: false
///
/// </summary>
///
/// Time: O(n)
/// Space: O(n)
///
/// We can use a recursive helper function to track the bounds of the left and right subtrees.
///
///
public class Solution {
    // Seems like a recursive helper function would be needed to track bounds

    public bool IsValidBST(TreeNode root) {
        return IsValid(root);
    }

    public bool IsValid(TreeNode node, int? min = null, int? max = null) {
        if (node == null) return true;

        if ((min.HasValue && node.val <= min.Value) ||
            (max.HasValue && node.val >= max.Value)) {
                return false;
        }

        // val is max for left and min for right

        return IsValid(node.left, min, node.val) &&
                IsValid(node.right, node.val, max);
    }
}
#endregion